cluster-autoscaler/simulator/basic_cluster_snapshot.go

```diff
@@ -0,0 +1,263 @@
+	havePodsWithAffinityList := make([]*schedulernodeinfo.NodeInfo, 0, len(data.nodeInfoMap))
+	for _, v := range data.nodeInfoMap {
+		if len(v.PodsWithAffinity()) > 0 {
+			havePodsWithAffinityList = append(havePodsWithAffinityList, v)
+		}
+	}
+	return havePodsWithAffinityList, nil
+}
+
+func (data *internalBasicSnapshotDataNodeLister) Get(nodeName string) (*schedulernodeinfo.NodeInfo, error) {
+	if v, ok := data.nodeInfoMap[nodeName]; ok {
+		return v, nil
+	}
+	return nil, fmt.Errorf("node %s not in snapshot", nodeName)
+}
+
+func (data *internalBasicSnapshotDataPodLister) List(selector labels.Selector) ([]*apiv1.Pod, error) {
+	alwaysTrue := func(p *apiv1.Pod) bool { return true }
+	return data.FilteredList(alwaysTrue, selector)
+}
+
+func (data *internalBasicSnapshotDataPodLister) FilteredList(podFilter schedulerlisters.PodFilter, selector labels.Selector) ([]*apiv1.Pod, error) {
+	pods := make([]*apiv1.Pod, 0)
+	for _, n := range data.nodeInfoMap {
+		for _, pod := range n.Pods() {
+			if podFilter(pod) && selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (data *internalBasicSnapshotData) Pods() schedulerlisters.PodLister {
+	return (*internalBasicSnapshotDataPodLister)(data)
+}
+
+func (data *internalBasicSnapshotData) NodeInfos() schedulerlisters.NodeInfoLister {
+	return (*internalBasicSnapshotDataNodeLister)(data)
+}
+
+// NewEmptySnapshot initializes a Snapshot struct and returns it.
+func newInternalBasicSnapshotData() *internalBasicSnapshotData {
+	return &internalBasicSnapshotData{
+		nodeInfoMap: make(map[string]*schedulernodeinfo.NodeInfo),
+	}
+}
+
+func (data *internalBasicSnapshotData) clone() *internalBasicSnapshotData {
+	clonedNodeInforMap := make(map[string]*schedulernodeinfo.NodeInfo)
+	for k, v := range data.nodeInfoMap {
+		clonedNodeInforMap[k] = v.Clone()
+	}
+	return &internalBasicSnapshotData{
+		nodeInfoMap: clonedNodeInforMap,
+	}
+}
+
+func (data *internalBasicSnapshotData) addNode(node *apiv1.Node) error {
+	if _, found := data.nodeInfoMap[node.Name]; found {
+		return fmt.Errorf("node %s already in snapshot", node.Name)
+	}
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	err := nodeInfo.SetNode(node)
+	if err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo; %v", err)
+	}
+	data.nodeInfoMap[node.Name] = nodeInfo
+	return nil
+}
+
+func (data *internalBasicSnapshotData) removeNode(nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		return fmt.Errorf("node %s not in snapshot", nodeName)
+	}
+	delete(data.nodeInfoMap, nodeName)
+	return nil
+}
+
+func (data *internalBasicSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		return fmt.Errorf("node %s not in snapshot", nodeName)
+	}
+	data.nodeInfoMap[nodeName].AddPod(pod)
+	return nil
+}
+
+func (data *internalBasicSnapshotData) removePod(namespace string, podName string) error {
```

Comment:
It's may be premature optimization, but I suspect it's quite likely you know the name of the node the pod is running on when deleting it. We could consider also adding RemovePodFromNode that wouldn't need to iterate over every pod in cluster (which seems like it can be expensive if done a lot in 5k node cluster).

Reply:
I think it may make sense. I will add a checkbox for this one.